### DISCUSSION TITLE
Authentication

### DISCUSSION THREAD

**Discussion by manson** [kamalshkeir/kago#11] [Ideas]
What is the best way to use an integrated authentication system in my app? I see parts of using it in different examples, but there is no special authentication workflow that is a kago way. I know, I can use User model, I can manually check passwords, I can build the whole chain of this, but I think I can miss some existing kago possibilities or overweight the solution. Is it possible to have a separate example of authentication process for a starting web-project to integrate it as fast as possible for a user? The common workflows like:
1) Registration. API - routes, standard functions to use. What parameters of income, what outcome. With email confirmation (with generating confirmation url) or without it as it is now.
1.1) Account activation route if email confirmation was used
2) Authorization. API - routes, standard functions to use. Tokens for user to result (or write to cookies or just return token).
3) Middleware auth and how to use it later - check if user logged in and access to a token's data or even update latest user object from database if necessary by just using a parameter for a middleware (for every call - ", refreshUserInfo") or else, probably right from the context (something like ctx.RefreshUserInfoFromDatabase()) .
4) Logout (drop tokens from cookies or else)

Repeat myself - I know I can pull out parts of this code from other examples (I saw it there) but... This is so common, usual and boring workflow that I (I'm sure not only me) would prefer just use it and not to spend time to investigate it every time or... Just use it, just bind to my web application and forget about it...

**Comment by kamalshkeir** [comment]
Hello bro, there is a common workflow for authentication, it is by using the middleware Auth, then you can access the user model and also IsAuthenticated which return bool from any handler https://github.com/kamalshkeir/kago/blob/master/core/kamux/context.go#L95 , https://github.com/kamalshkeir/kago/blob/master/core/kamux/context.go#L104, and from any html page like so https://github.com/kamalshkeir/kago-assets/blob/master/templates/admin/admin_nav.html#L17, this middleware store the session token cookie, which is an AES token , and finaly for the logout, you only need to delete cookie session, using for example https://github.com/kamalshkeir/kago-assets/blob/master/templates/admin/admin_nav.html#L17

**Reply by kamalshkeir** [reply]
you can use middleware Auth and Admin like so https://github.com/kamalshkeir/kago/blob/master/core/admin/urls.go#L19

**Reply by manson** [reply]
Nice, thanks. I'll just need to combine all this, that is related to authentication, all links you supplied into a usable Sheet Cheat :-)
